content/courses/advanced-ada/parts/resource_management/access_types.rst

```diff
@@ -3445,26 +3445,401 @@ caller.
-    a bank account that involves dynamic allocation for some part of an
-    account, the bank account is a private type (or limited), and the access
-    values are hidden in the private part. Thus when an account is closed the
-    allocated memory can be freed without the possibility of leaving dangling
-    references. That will work as long as the reference cannot be copied when
-    copying bank account objects, in which case we'd want the type to be
-    limited as well as private. Or we'd make it a controlled type so that
-    Finalize can call Free.
+    Add link to Adv_Ada_Finalization
+
+Let's adapt the previous example and declare :ada:`Info` as a limited
+controlled type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Access_Types.Design_Strategies.Access_Type_Limited_Controlled_Abstraction
+
+    with Ada.Finalization;
+
+    package Access_Type_Abstraction is
+
+       type Info is limited private;
+
+       function To_Info (S : String) return Info;
+
+       function To_String (Obj : Info)
+                           return String;
+
+       function Copy (Obj : Info) return Info;
+
+       procedure Copy (To   : in out Info;
+                       From :        Info);
+
+       procedure Append (Obj : in out Info;
+                         S   :        String);
+
+       procedure Reset (Obj : in out Info);
+
+    private
+
+       type String_Access is access String;
+
+       type Info is new
+         Ada.Finalization.Limited_Controlled with
+          record
+             Str_A : String_Access;
+          end record;
+
+       procedure Initialize (Obj : in out Info);
+       procedure Finalize (Obj : in out Info);
+
+    end Access_Type_Abstraction;
+
+    with Ada.Unchecked_Deallocation;
+
+    package body Access_Type_Abstraction is
+
+       --
+       --  STRING_ACCESS SUBPROGRAMS
+       --
+
+       function To_String_Access (S : String)
+                                  return String_Access
+       is
+         (new String'(S));
+
+       function To_String (S : String_Access)
+                           return String is
+         (if S /= null then S.all else "");
+
+       procedure Free is
+         new Ada.Unchecked_Deallocation
+           (Object => String,
+            Name   => String_Access);
+
+       procedure Copy (To   : in out String_Access;
+                       From :        String_Access) is
+       begin
+          Free (To);
+          To := To_String_Access (From.all);
+       end Copy;
+
+       procedure Append (Obj : in out String_Access;
+                         S   :        String) is
+          New_Str : constant String_Access :=
+                      To_String_Access
+                        (To_String (Obj) & S);
+       begin
+          Free (Obj);
+          Obj := New_Str;
+       end Append;
+
+       --
+       --  PRIVATE SUBPROGRAMS
+       --
+
+       procedure Initialize (Obj : in out Info) is
+       begin
+          Obj.Str_A := null;
```

Comment:
Since null is the component's default value anyway, why do we need procedure Initialize?

Reply:
You're right, this isn't strictly needed. However, I'd like to keep this line in order to illustrate the automatic calls to  `Initialize` and `Finalize` for controlled types in this context. I'll add a big note to the procedure explaining that this assignment is actually useless in "production" code. I hope this is OK with you.

Reply:
Also, I'll add a call to `Put_Line` in the comments.

Reply:
Anyway, if you prefer, I could replace that assignment by a `null;`.

Reply:
A note about it not being necessary for that specific component type would be a good idea, but otherwise I've no objection to keeping it.